Default markdown template (#483)
twisted/towncrier: docs/configuration.rst

=======================

``towncrier`` has many knobs and switches you can use, to customize it to your project's needs.

> The setup in the `Tutorial <tutorial.html>`_ doesn't touch on many, but this document will detail each of these options for you!

Reviewer: the hell are these links with html suffixes – mind changing them to proper Sphinx links? In this case it should be enough to use
```rst
:doc:`tutorial`
```

With customization below you may want to add an explicit name if you want to, but I think

```rst
:doc:`customization`
```

should be fine

For how to perform common customization tasks, see `Customization <customization/index.html>`_.

``[tool.towncrier]``
--------------------

All configuration for ``towncrier`` sits inside ``towncrier.toml`` or ``pyproject.toml``, under the ``tool.towncrier`` namespace.
Please see https://toml.io/ for how to write TOML.

A minimal configuration for a Python project looks like this:

.. code-block:: toml

   # pyproject.toml

   [tool.towncrier]
   package = "myproject"

A minimal configuration for a non-Python project looks like this:

.. code-block:: toml

   # towncrier.toml

   [tool.towncrier]
   name = "My Project"

>    version = "1.0.0"

Reviewer: Maybe worth it to mention that the version is optional & can be passed on command line? To me it doesn't make much sense to keep the project version separately in a config file.

Author: This says "minimal configuration", so I'll just take it out :P

   directory = "newsfragments"

Top level keys
~~~~~~~~~~~~~~

- **``name``** -- The name of your project. If empty and the ``package`` key is provided, the name will be automatically determined.
  ``""`` by default.

> - **``version``** -- The version of your project. Mandatory except for Python projects that provide the ``package`` key and want the version to be automatically determined from a ``__version__`` variable in the package's module.

Reviewer: It's not really mandatory – you just have to pass the version explicitly on the command-line as [we do in attrs](https://github.com/python-attrs/attrs/blob/56cbf826df6adf04dd176fdd5391cf8829a14aa5/tox.ini#L52) for instance.

Reviewer: ```suggestion
- **``version``** -- The version of your project.
  Python projects that provide the ``package`` key can have the version to be automatically determined from a ``__version__`` variable in the package's module.
  The version can also be passed explicitly by command line argument ``--version``.
```

> - **``directory``** -- The directory storing you news fragments. Mandatory except for Python projects (where the default is a ``newsfragments`` directory within the package).

Reviewer: ```suggestion
- **``directory``** -- The directory storing your news fragments. Mandatory except for Python projects (where the default is a ``newsfragments`` directory within the package).
```

> - **``directory``** -- The directory storing you news fragments. Mandatory except for Python projects (where the default is a ``newsfragments`` directory within the package).

Reviewer: ```suggestion
- **``directory``** -- The directory storing you news fragments.
  Mandatory except for Python projects (where the default is a ``newsfragments`` directory within the package).
```

- **``filename``** -- The filename of your news file.
  ``"NEWS.rst"`` by default.

> - **``template``** -- Path to the template for generating the news file. If the path starts with ``towncrier:``, it is interpreted as a template bundled with ``towncrier``.

Reviewer: ```suggestion
- **``template``** -- Path to the template for generating the news file.
  If the path starts with ``towncrier:``, it is interpreted as a template bundled with ``towncrier``.
```

  ``"towncrier:default.rst"`` by default (unless ``filename`` ends with ``.md``, in which case the default is ``"towncrier:default.md"``).
- **``start_string``** -- The magic string that ``towncrier`` looks for when considering where the release notes should start.
  ``".. towncrier release notes start"`` by default.

> - **``title_format``** -- A format string for the title of your project. The explicit value of ``False`` will disable the title entirely. Any other empty value means the template should render the title (the bundled templates use ``<name> <version> (<date>)``). Strings should use the following keys to render the title dynamically: ``{name}``, ``{version}``, and ``{project_date}``.

Reviewer: ```suggestion
- **``title_format``** -- A format string for the title of your project.
  The explicit value of ``False`` will disable the title entirely.
  Any other empty value means the template should render the title (the bundled templates use ``<name> <version> (<date>)``).
  Strings should use the following keys to render the title dynamically: ``{name}``, ``{version}``, and ``{project_date}``.
```

  ``""`` by default.

> - **``issue_format``** -- A format string for rendering the issue/ticket number in newsfiles. If none, the issues are rendered as ``#<issue>`` if for issues that are integers, or just ``<issue>`` otherwise. Use the ``{issue}`` key in your string render the issue number, for example Markdown projects may want to use ``"[{issue}]: https://<your bug tracker>/{issue}"``.

Reviewer: ```suggestion
- **``issue_format``** -- A format string for rendering the issue/ticket number in newsfiles.
  If none, the issues are rendered as ``#<issue>`` if for issues that are integers, or just ``<issue>`` otherwise.
  Use the ``{issue}`` key in your string render the issue number, for example Markdown projects may want to use ``"[{issue}]: https://<your bug tracker>/{issue}"``.
```

  ``None`` by default.

> - **``underlines``** -- The characters used for underlining headers. Not used in the bundled Markdown template.

Reviewer: ```suggestion
- **``underlines``** -- The characters used for underlining headers.
  Not used in the bundled Markdown template.
```

  ``["=", "-", "~"]`` by default.

> - **``wrap``** -- Boolean value indicating whether to wrap news fragments to a width of 79.

Reviewer: ```suggestion
- **``wrap``** -- Boolean value indicating whether to wrap news fragments to a line length of 79.
```

  ``false`` by default.
- **``all_bullets``** -- Boolean value indicating whether the template uses bullets for each news fragment.
  ``true`` by default.

> - **``single_file``** -- Boolean value indicating whether to write all news fragments to a single file. If false, the ``filename`` should use the following keys to render the filenames dynamically: ``{name}``, ``{version}``, and ``{project_date}``.

Reviewer: ```suggestion
- **``single_file``** -- Boolean value indicating whether to write all news fragments to a single file.
  If false, the ``filename`` should use the following keys to render the filenames dynamically:
  ``{name}``, ``{version}``, and ``{project_date}``.
```

  ``true`` by default.
- **``orphan_prefix``** -- The prefix used for orphaned news fragments.
  ``"+"`` by default.

Extra top level keys for Python projects
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

- **``package``** -- The package name of your project.
- **``package_dir``** -- The folder your package lives.
  ``"."`` by default, some projects might need to use ``"src"``.

Sections
--------

``towncrier`` supports splitting fragments into multiple sections, each with its own news of fragment types.

Add an array of tables your ``.toml`` configuration file named **``[[tool.towncrier.section]]``**.

Each table within this array has the following mandatory keys:

- **``name``** -- The name of the section.

> - **``path``** -- The path to the directory containing the news fragments for this section, relative to the configured ``directory``. Use ``""`` for the root directory.

Reviewer: ```suggestion
- **``path``** -- The path to the directory containing the news fragments for this section, relative to the configured ``directory``.
  Use ``""`` for the root directory.
```

For example:

.. code-block:: toml

   [[tool.towncrier.section]]
   name = "Main Platform"
   path = ""

   [[tool.towncrier.section]]
   name = "Secondary"
   path = "secondary"

Custom fragment types
---------------------

> ``towncrier`` hase the following default fragment types: ``feature``, ``bugfix``, ``doc``, ``removal``, and ``misc``.

Reviewer: ```suggestion
``towncrier`` has the following default fragment types: ``feature``, ``bugfix``, ``doc``, ``removal``, and ``misc``.
```
